### PR TITLE
Blog post for Unified Theory of Garbage Collection (11/04)

### PR DESCRIPTION
**Pull request by mdmoeller**
Merging my blog post for Wed. of next week!

### CODE REVIEW

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+There are two classical algorithms for automatic garbage collection, tracing
+(also called mark and sweep), and reference counting, which were both first
+developed in the 1960s.  [Java's generational garbage collector][java] is a
+typical example of a tracing collector, and [CPython's garbage
+collector][python] uses reference counting---but as this paper suggests, the
+reality is that these distinctions are certainly muddier in practice.
+
+The two paradigms are associated important performance tradeoffs. In particular,
+the mark and sweep approach adds significant pauses to program progress, since
+in the simplest approach one must pause execution to 'trace' all live objects
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md]: I'd probably use double quotes at the "top level" of nesting.

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+incrementally. Reference counting's major drawback is the difficulty presented
+by cycles appearing the link structure, which prevent objects' reference counts
+from ever reaching 0.
+
+## Highlights
+
+Bacon et al. begin by presenting the two fundamental garbage collection
+algorithms, reference counting and tracing, as dual views of the same underlying
+structure. This should hit you a bit like hearing that butter and I Can't
+Believe It's Not Butter have the same deep structure.
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md:39]: 😂

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  "nursery" to improve the time spent on collection. The idea is that recently
+  allocated objects are more likely to be freed earlier, so we can spend our
+  garbage-collection time more efficiently by looking at newer objects more
+  often.
+
+* **Older First.** (Sketched only). In this approach, the oldest generation
+  is traced first instead of the nursery.
+
+* **GCH: Generational with Counted Heap.** Run normal tracing of the nursery,
+  but do reference counting on the mature remaining space of the heap.
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md]: Neither here nor there, but: this always struck me as a pretty sweet design point if you can make it work.

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+* **Train Algorithm.** (Presented but not analyzed). While this approach was
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md]:
```suggestion
* **Train Algorithm.** (Presented but not analyzed.) While this approach was
```

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  definitely interesting, and they drive home that it fits into their hybrid
+  model (car ordering provides implicit reference counts), I must admit I was not
+  able to glean much intuition about why the approach would be worth
+  implementing.
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md]: Maybe this bullet could use one more sentence about what the approach actually is before adding this additional commentary?

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+* **CC: Reference Counting, Trial Deletion.** This is mostly reference counting,
+  but with some tracing of certain "candidates" mixed in to detect cycles (i.e.
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md]:
```suggestion
  but with some tracing of certain "candidates" mixed in to detect cycles (i.e.,
```
https://capra.cs.cornell.edu/styleguide/#egie

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+* **C: Reference Counting.** This is the classic reference counting algorithm. 
+  They give a really slick way to reduce space
+  overhead for traversing the links by using the space guaranteed to exist in
+  the dead objects themselves.
+
+The paper concludes with a mathematical analysis of the costs of the
+above collection strategies. The authors point out that the analyses are not 
+big-O, so these costs are actually "real", up to the accuracy of the constants
+for a particular system (and I guess also the constancy of those constants?).
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md]: Hmm, not entirely sure what you mean by this parenthetical—are you suggesting that it's not entirely certain that these factors are actually constants in all settings? That would be reasonable!

File: content/blog/2020-11-04-unified-theory-gc/index.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+organization is quite nice.  That is, they give the deep comparison abstractly
+in several different ways, then completely flesh out how existing hybrids
+combine the two flavors.
+
+The section of the paper that has the most potentional to be practically
+important is the cost analysis in section 7.  Having never constructed a garbage
+collection system for a whole language, I do not know whether a
+compiler designer would actually use these formulas to tune a garbage collector
+(even with knowledge of their particular domain) since they would have to
+validate any decisions against benchmark performance anyway.
```
Comment (sampsyo) [inline content/blog/2020-11-04-unified-theory-gc/index.md:155]: Great question; I don't know either! I can see it perhaps being useful as a first step, in combination with validation on real benchmarks.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Awesome work! I thought the list of brief summaries of the different intermediate design points was very clear and a good way to structure the post.

I have just a couple of nitpicky suggestions within.

**Comment by mdmoeller** [comment]
Incorporated all suggested changes into 17a0b7b.

**Review by sampsyo — APPROVED** [review]
Excellent.

**Pull request merged**